File: downloader/lib/outstand.py
"""Minimal Outstand.so API client — ports the Instagram-relevant calls.

Mirrors the TypeScript client in ../socialMedia/reelGen/lib/outstand.ts.
Only the methods we need for: listing IG accounts, uploading a video, and
creating a post are ported.
"""

import os
from pathlib import Path
from typing import Any

import requests

DEFAULT_BASE_URL = "https://api.outstand.so/v1"


class OutstandError(RuntimeError):
    pass


def _base_url() -> str:
    return (os.environ.get("OUTSTAND_BASE_URL") or DEFAULT_BASE_URL).rstrip("/")


def _api_key() -> str:
    key = (os.environ.get("OUTSTAND_API_KEY") or "").strip()
    if not key:
        raise OutstandError("OUTSTAND_API_KEY is not set — check .env")
    return key


def _headers(auth: bool = True, json_body: bool = False) -> dict[str, str]:
    h = {"Accept": "application/json"}
    if auth:
        h["Authorization"] = f"Bearer {_api_key()}"
    if json_body:
        h["Content-Type"] = "application/json"
    return h


def _unwrap(payload: Any) -> Any:
    """Outstand wraps responses as {success, data, ...}. Unwrap or raise."""
    if isinstance(payload, dict) and isinstance(payload.get("success"), bool):
        if not payload["success"]:
            raise OutstandError(payload.get("message") or payload.get("error") or "Outstand request failed")
        if "data" in payload:
            return payload["data"]
    return payload


def _request(method: str, path: str, *, auth: bool = True, query: dict | None = None, json: Any = None) -> Any:
    url = f"{_base_url()}/{path.lstrip('/')}"
    r = requests.request(
        method=method,
        url=url,
        headers=_headers(auth=auth, json_body=json is not None),
        params=query,
        json=json,
        timeout=60,
    )
    if not r.ok:
        raise OutstandError(f"Outstand {method} {path} → {r.status_code}: {r.text[:300]}")
    if not r.text:
        return {}
    try:
        return _unwrap(r.json())
    except ValueError:
        return {}


# ---------- accounts ----------

def list_social_accounts(limit: int = 100, offset: int = 0) -> list[dict]:
    payload = _request("GET", "/social-accounts", query={"limit": limit, "offset": offset})
    if isinstance(payload, list):
        return payload
    if isinstance(payload, dict):
        for key in ("accounts", "social_accounts", "items", "data"):
            val = payload.get(key)
            if isinstance(val, list):
                return val
    return []


def list_instagram_accounts() -> list[dict]:
    return [
        a for a in list_social_accounts()
        if str(a.get("network") or a.get("platform") or "").lower() == "instagram"
    ]


# ---------- media upload ----------

def _pick_filename(content_type: str, base: str) -> str:
    ct = (content_type or "").lower()
    if "quicktime" in ct:
        return f"{base}.mov"
    if "webm" in ct:
        return f"{base}.webm"
    return f"{base}.mp4"


def upload_local_video(local_path: Path, filename_base: str) -> dict:
    """Upload a local mp4 to Outstand. Returns {media_id, url, filename}."""
    local_path = Path(local_path)
    if not local_path.exists():
        raise OutstandError(f"video not found: {local_path}")

    content_type = "video/mp4"
    filename = _pick_filename(content_type, filename_base)

    # 1. Request upload URL
    create = _request(
        "POST", "/media/upload",
        json={"filename": filename, "content_type": content_type},
    )
    media_id = create.get("media_id") or create.get("mediaId") or create.get("id")
    upload_url = create.get("upload_url") or create.get("uploadUrl")
    if not media_id or not upload_url:
        raise OutstandError(f"Outstand upload URL missing: {create}")

    # 2. PUT the file bytes
    with open(local_path, "rb") as f:
        put = requests.put(
            upload_url,
            data=f,
            headers={"Content-Type": content_type},
            timeout=300,
        )
    if not put.ok:
        raise OutstandError(f"Outstand PUT failed ({put.status_code}): {put.text[:300]}")

    # 3. Confirm
    confirmed = _request("POST", f"/media/{media_id}/confirm")
    media = confirmed.get("media") if isinstance(confirmed, dict) and isinstance(confirmed.get("media"), dict) else confirmed
    media = media if isinstance(media, dict) else {}
    url = media.get("url") or create.get("url")
    if not url:
        raise OutstandError("Outstand confirm returned no media url")

    return {"media_id": media_id, "url": url, "filename": filename}


# ---------- posts ----------

def create_post(accounts: list[str], caption: str, media_url: str, filename: str) -> dict:
    payload = _request(
        "POST", "/posts/",
        json={
            "accounts": accounts,
            "content": caption,
            "containers": [
                {
                    "platform": "instagram",
                    "content": caption,
                    "media": [{"url": media_url, "filename": filename}],
                }
            ],
        },
    )
    post = payload.get("post") if isinstance(payload, dict) and isinstance(payload.get("post"), dict) else payload
    post = post if isinstance(post, dict) else {}
    post_id = post.get("id") or post.get("post_id") or post.get("postId")
    if not post_id:
        raise OutstandError(f"Outstand post id missing: {payload}")
    return {
        "post_id": post_id,
        "status": post.get("status") or post.get("state") or "submitted",
    }


def get_post_status(post_id: str) -> dict:
    payload = _request("GET", f"/posts/{post_id}/")
    post = payload.get("post") if isinstance(payload, dict) and isinstance(payload.get("post"), dict) else payload
    post = post if isinstance(post, dict) else {}
    return {
        "post_id": post.get("id") or post.get("post_id") or post_id,
        "status": post.get("status") or post.get("state"),
        "external_id": post.get("external_id") or post.get("externalId"),
        "error": post.get("error"),
    }

File: lib/outstand.ts
import "server-only";

const DEFAULT_BASE_URL = "https://api.outstand.so/v1";

export type OutstandAccount = {
  id: string;
  network: string;
  username?: string;
  displayName?: string;
  avatarUrl?: string;
  raw: Record<string, unknown>;
};

export type OutstandMedia = {
  mediaId: string;
  url: string;
  filename: string;
};

function baseUrl() {
  return (process.env.OUTSTAND_BASE_URL || DEFAULT_BASE_URL).replace(/\/+$/, "");
}

function apiKey() {
  const k = (process.env.OUTSTAND_API_KEY || "").trim();
  if (!k) throw new Error("OUTSTAND_API_KEY is not set");
  return k;
}

function unwrap<T = unknown>(payload: unknown): T {
  if (payload && typeof payload === "object" && "success" in (payload as Record<string, unknown>)) {
    const p = payload as { success: boolean; data?: unknown; message?: string; error?: string };
    if (p.success === false) {
      throw new Error(p.message || p.error || "Outstand request failed");
    }
    if ("data" in p && p.data !== undefined) return p.data as T;
  }
  return payload as T;
}

async function request<T>(method: "GET" | "POST" | "PUT", path: string, init?: { query?: Record<string, string | number | undefined>; body?: unknown; authenticated?: boolean }): Promise<T> {
  const url = new URL(`${baseUrl()}/${path.replace(/^\/+/, "")}`);
  if (init?.query) {
    for (const [k, v] of Object.entries(init.query)) {
      if (v !== undefined) url.searchParams.set(k, String(v));
    }
  }
  const headers: Record<string, string> = { Accept: "application/json" };
  if (init?.authenticated !== false) headers["Authorization"] = `Bearer ${apiKey()}`;
  if (init?.body !== undefined) headers["Content-Type"] = "application/json";

  const res = await fetch(url, {
    method,
    headers,
    body: init?.body !== undefined ? JSON.stringify(init.body) : undefined,
  });
  const text = await res.text();
  if (!res.ok) throw new Error(`Outstand ${method} ${path} → ${res.status}: ${text.slice(0, 300)}`);
  if (!text) return {} as T;
  try {
    return unwrap<T>(JSON.parse(text));
  } catch {
    return {} as T;
  }
}

function str(...vals: unknown[]): string | undefined {
  for (const v of vals) if (typeof v === "string" && v.trim()) return v.trim();
  return undefined;
}

function normalizeAccount(raw: Record<string, unknown>): OutstandAccount {
  const id = str(raw.id, raw.account_id, raw.outstand_account_id);
  const network = str(raw.network, raw.platform);
  if (!id || !network) throw new Error("Outstand account missing id/network");
  return {
    id,
    network,
    username: str(raw.username),
    displayName: str(raw.display_name, (raw as Record<string, unknown>).displayName, raw.name),
    avatarUrl: str(raw.avatar_url, (raw as Record<string, unknown>).avatarUrl),
    raw,
  };
}

// ---------- accounts ----------

export async function listInstagramAccounts(): Promise<OutstandAccount[]> {
  const payload = await request<unknown>("GET", "/social-accounts", { query: { limit: 100, offset: 0 } });
  const list: unknown[] = Array.isArray(payload)
    ? payload
    : (payload && typeof payload === "object")
      ? (() => {
          const p = payload as Record<string, unknown>;
          for (const key of ["accounts", "social_accounts", "items", "data"]) {
            if (Array.isArray(p[key])) return p[key] as unknown[];
          }
          return [];
        })()
      : [];

  return list
    .filter((a): a is Record<string, unknown> => !!a && typeof a === "object")
    .map(normalizeAccount)
    .filter((a) => a.network.toLowerCase() === "instagram");
}

// ---------- media upload ----------

function pickFilename(contentType: string, base: string): string {
  const ct = contentType.toLowerCase();
  if (ct.includes("quicktime")) return `${base}.mov`;
  if (ct.includes("webm")) return `${base}.webm`;
  return `${base}.mp4`;
}

/** Download the remote URL and upload it to Outstand, returning the canonical media url. */
export async function uploadRemoteVideo(input: { sourceUrl: string; filenameBase: string }): Promise<OutstandMedia> {
  const res = await fetch(input.sourceUrl);
  if (!res.ok) throw new Error(`Source video fetch failed (${res.status}) ${input.sourceUrl}`);
  const contentType = res.headers.get("content-type") || "video/mp4";
  const filename = pickFilename(contentType, input.filenameBase);

  const createPayload = await request<Record<string, unknown>>("POST", "/media/upload", {
    body: { filename, content_type: contentType },
  });
  const mediaId = str(createPayload.media_id, createPayload.mediaId, createPayload.id);
  const uploadUrl = str(createPayload.upload_url, createPayload.uploadUrl);
  if (!mediaId || !uploadUrl) throw new Error("Outstand upload URL missing from response");

  const put = await fetch(uploadUrl, {
    method: "PUT",
    headers: { "Content-Type": contentType },
    body: await res.arrayBuffer(),
  });
  if (!put.ok) {
    const text = await put.text();
    throw new Error(`Outstand PUT failed (${put.status}): ${text.slice(0, 300)}`);
  }

  const confirmPayload = await request<Record<string, unknown>>("POST", `/media/${encodeURIComponent(mediaId)}/confirm`);
  const media = (confirmPayload.media && typeof confirmPayload.media === "object")
    ? (confirmPayload.media as Record<string, unknown>)
    : confirmPayload;
  const url = str((media as Record<string, unknown>).url) || str(createPayload.url);
  if (!url) throw new Error("Outstand confirm missing media url");

  return { mediaId, url, filename };
}

// ---------- posts ----------

export type OutstandPostResult = { postId: string; status: string };

export async function createPost(input: {
  accounts: string[];
  caption: string;
  mediaUrl: string;
  filename: string;
}): Promise<OutstandPostResult> {
  const payload = await request<Record<string, unknown>>("POST", "/posts/", {
    body: {
      accounts: input.accounts,
      content: input.caption,
      containers: [
        {
          platform: "instagram",
          content: input.caption,
          media: [{ url: input.mediaUrl, filename: input.filename }],
        },
      ],
    },
  });
  const post = (payload.post && typeof payload.post === "object")
    ? (payload.post as Record<string, unknown>)
    : payload;
  const postId = str((post as Record<string, unknown>).id, (post as Record<string, unknown>).post_id, (post as Record<string, unknown>).postId);
  if (!postId) throw new Error(`Outstand post id missing: ${JSON.stringify(payload).slice(0, 200)}`);
  return {
    postId,
    status: str((post as Record<string, unknown>).status, (post as Record<string, unknown>).state) || "submitted",
  };
}

export async function getPostStatus(postId: string): Promise<{ status: string; externalId?: string; error?: string }> {
  const payload = await request<Record<string, unknown>>("GET", `/posts/${encodeURIComponent(postId)}/`);
  const post = (payload.post && typeof payload.post === "object")
    ? (payload.post as Record<string, unknown>)
    : payload;
  return {
    status: str((post as Record<string, unknown>).status, (post as Record<string, unknown>).state) || "",
    externalId: str((post as Record<string, unknown>).external_id, (post as Record<string, unknown>).externalId),
    error: str((post as Record<string, unknown>).error),
  };
}
